src/libstore/derivations.cc

```diff
@@ -129,9 +129,11 @@ static DerivationOutput parseDerivationOutput(const Store & store, istringstream
 }
 
 
-static Derivation parseDerivation(const Store & store, const string & s)
+static Derivation parseDerivation(const Store & store, const string & s, string name)
```

Comment:
Shouldn't we rather read the name from `drv.env["name"]` to make them more self-contained ? Atm it's technically possible to realize a derivation that doesn't have a `name` field (or has a `name` field that's different from the name of the store path), but there's no way to build such a derivation (even manually because it won't be possible to import it into the store)

Reply:
@regnat I would consider magic things in env a libexpr convention rather than it's nice for libstore to not be aware of.

Matt and I talked about getting the name from the derivation output paths during parsing, which would use libstore-aware conventions. We didn't do that in this PR cause we could get away with it, but if we need to get rid of this new argument we can do that.

Reply:
> I would consider magic things in env a libexpr convention rather than it's nice for libstore to not be aware of.

I agree in principle, but given that the drv format is not extensible that's the best way we have to pass meta information about a derivation. And there's quite a bit of precedent about reading information from the `env` field (`preferLocalBuild`, `__json`, `requiredSystemFeatures`…), so it seems legit to also do that for the name